web-server-lib/web-server/dispatchers/dispatch-logresp.rkt

```diff
@@ -0,0 +1,127 @@
+(define log-format/c (symbols 'parenthesized-default 'extended 'apache-default))
+
+(provide/contract
+ [format-reqresp/c contract?]
+ [log-format/c contract?]
+ [log-format->format (log-format/c . -> . format-reqresp/c)]
+ [paren-format format-reqresp/c]
+ [extended-format format-reqresp/c]
+ [apache-default-format format-reqresp/c]
+ [interface-version dispatcher-interface-version/c]
+ [make (->* (dispatcher/c)
+            (#:format (or/c log-format/c format-reqresp/c)
+             #:log-path (or/c path-string? output-port?))
+            dispatcher/c)])
+
+(define ((log-header-handler log-message req original-handler) resp)
+  (log-message req resp)
+  (original-handler resp))
+
+(define interface-version 'v1)
+(define (make #:format [format paren-format]
+              #:log-path [log-path "log"]
+              dispatcher)
+  (define final-format
+    (if (symbol? format)
+        (log-format->format format)
+        format))
+  (define log-message (make-log-message log-path final-format))
+  (lambda (conn req)
+    (with-handlers ([exn:dispatcher? (lambda (e) (next-dispatcher))])
+      (parameterize ([current-header-handler (log-header-handler log-message req (current-header-handler))])
+        (dispatcher conn req)))))
+
+(define (log-format->format log-format)
+  (case log-format
+    [(parenthesized-default)
+     paren-format]
+    [(extended)
+     extended-format]
+    [(apache-default)
+     apache-default-format]))
+
+(define (request-line-raw req)
+  (format "~a ~a HTTP/1.1"
+          (string-upcase (bytes->string/utf-8 (request-method req)))
+          (url->string (request-uri req))))
+(define (apache-default-format req resp)
+  (define request-time (srfi-date:current-date))
+  (format "~a - - [~a] \"~a\" ~a ~a\n"
+          (request-client-ip req)
+          (srfi-date:date->string request-time "~d/~b/~Y:~T ~z")
+          (request-line-raw req)
+          (response-code resp)
+          "-"))
+
+(define (paren-format req resp)
+  (format "~s\n"
+          (list 'from (request-client-ip req)
+                'to (request-host-ip req)
+                'for (url->string (request-uri req))
+                'at (date->string (seconds->date (current-seconds)) #t)
+                'code (response-code resp))))
+
+(define (extended-format req resp)
+  (format "~s\n"
+          `((client-ip ,(request-client-ip req))
+            (host-ip ,(request-host-ip req))
+            (referer ,(let ([R (headers-assq* #"Referer" (request-headers/raw req))])
+                        (if R
+                            (header-value R)
+                            #f)))
+            (uri ,(url->string (request-uri req)))
+            (time ,(current-seconds))
+            (code ,(response-code resp)))))
+
+(define (make-log-message log-path-or-port format-reqresp)
+  (define log-ch (make-async-channel))
+  (define log-thread
+    (thread/suspend-to-kill
+     (lambda ()
+       (let loop ([log-p #f])
+         (sync
```

Comment:
If you're not planning to handle other events here, you could drop the `sync`+`handle-evt` bits and just call `async-channel-get`. Additionally, you could use the thread's built-in mailbox (`thread-send`, `thread-receive`) instead of an async channel.

Also, note that using a buffered channel (whether it's an async channel or the thread's mailbox) here means that pending logs can pile up if disk IO is slow. If the request rate is greater than the rate at which things can be logged, these queues will grow without bound (Little's law). I would use a regular channel instead to create back pressure and to make the cost of using the dispatcher clear.

Reply:
FYI: this code was copied from https://github.com/racket/web-server/blob/master/web-server-lib/web-server/dispatchers/dispatch-log.rkt#L84. If I have to write it from scratch, I would not have written it this way either. 

For now, I think I will avoid touching this bit of code, since I'm not confident if there is any subtle issue that the code is handling. However, feel free to submit a PR to improve it afterwards. Alternatively, you can submit a PR to improve `dispatch-log.rkt` now -- and I will reflect the change to `dispatch-logresp.rkt`.

Reply:
Can you NOT the copy the code, but abstract it?

Reply:
@jeapostrophe I will wait until #120 is merged first, since this PR and that PR both touch the same file, and I don't want to create a conflict.

Reply:
Done.